packages/ics721/src/execute.rs

```diff
@@ -336,32 +350,35 @@ where
         &self,
         deps: DepsMut,
         env: Env,
-        info: MessageInfo,
+        nft_contract: &Addr,
         token_id: TokenId,
         nft_owner: String,
         msg: Binary,
     ) -> Result<Response<T>, ContractError> {
         let nft_owner = deps.api.addr_validate(&nft_owner)?;
         let msg: IbcOutgoingMsg = from_json(msg)?;
 
-        let class = match NFT_CONTRACT_TO_CLASS_ID.may_load(deps.storage, info.sender.clone())? {
+        let class = match load_class_id_for_nft_contract(deps.as_ref().storage, nft_contract)? {
             Some(class_id) => CLASS_ID_TO_CLASS.load(deps.storage, class_id)?,
             // No class ID being present means that this is a local NFT
             // that has never been sent out of this contract.
             None => {
-                let class_data = self.get_class_data(&deps, &info.sender)?;
+                let class_data = self.get_class_data(&deps, nft_contract)?;
                 let data = class_data.as_ref().map(to_json_binary).transpose()?;
                 let class = Class {
-                    id: ClassId::new(info.sender.to_string()),
+                    id: ClassId::new(nft_contract.to_string()),
                     // There is no collection-level uri nor data in the
                     // cw721 specification so we set those values to
                     // `None` for local, cw721 NFTs.
                     uri: None,
                     data,
                 };
 
-                NFT_CONTRACT_TO_CLASS_ID.save(deps.storage, info.sender.clone(), &class.id)?;
-                CLASS_ID_TO_NFT_CONTRACT.save(deps.storage, class.id.clone(), &info.sender)?;
```

Comment:
@humanalgorithm as noted in audit:

> In the cw-ics721 package, specifically in state.rs at lines 22 and 25, there are two cw-storage-plus::Maps
named CLASS_ID_TO_NFT_CONTRACT and NFT_CONTRACT_TO_CLASS_ID. These maps are responsible for tracking the class_id and collection_addr in a reciprocal manner. However, this setup leads to redundant operations since any modification to these values necessitates updates in both maps.

So both maps hold redundant data - which requires redundant operations like here.

Reply:
That's funny, I originally proposed the IndexedMap in the original implementation but my suggestion got shot down.